src/deadlock-detection/threaddeadlock.cpp

```diff
@@ -0,0 +1,663 @@
+        }
+    }
+}
+
+void SetWaitingToHeld(void *c, OwnershipType ownership)
+{
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+
+    const uint64_t tid = getTid();
+    // we do this first so changes are still made for trylocks which dont have
+    // a waiting lock to be edited
+    auto itheld = lockdata.locksheldbythread.find(tid);
+    if (itheld != lockdata.locksheldbythread.end())
+    {
+        for (auto rit = itheld->second.rbegin(); rit != itheld->second.rend(); ++rit)
+        {
+            if (rit->first == c)
+            {
+                rit->second.ChangeWaitingToHeld();
+                break;
+            }
+        }
+    }
+
+
+    if (ownership == OwnershipType::EXCLUSIVE)
+    {
+        auto it = lockdata.writelockswaiting.find(c);
+        if (it != lockdata.writelockswaiting.end())
+        {
+            it->second.erase(tid);
+        }
+        else
+        {
+            DbgAssert(!"Missing write lock waiting", );
+        }
+        auto iter = lockdata.writelocksheld.find(c);
+        if (iter == lockdata.writelocksheld.end())
+        {
+            std::set<uint64_t> holders;
+            holders.emplace(tid);
+            lockdata.writelocksheld.emplace(c, holders);
+        }
+        else
+        {
+            iter->second.emplace(tid);
+        }
+    }
+    else //  !isExclusive
+    {
+        auto it = lockdata.readlockswaiting.find(c);
+        if (it != lockdata.readlockswaiting.end())
+        {
+            it->second.erase(tid);
+        }
+        else
+        {
+            DbgAssert(!"Missing read lock waiting", );
+        }
+        auto iter = lockdata.readlocksheld.find(c);
+        if (iter == lockdata.readlocksheld.end())
+        {
+            std::set<uint64_t> holders;
+            holders.emplace(tid);
+            lockdata.readlocksheld.emplace(c, holders);
+        }
+        else
+        {
+            iter->second.emplace(tid);
+        }
+    }
+}
+
+// c = the cs
+// isExclusive = is the current lock exclusive, for a recursive mutex (CCriticalSection) this value should always be
+// true
+void push_lock(void *c, const CLockLocation &locklocation, LockType locktype, OwnershipType ownership, bool fTry)
+{
+    std::lock_guard<std::mutex> lock(lockdata.dd_mutex);
+
+    LockStackEntry now = std::make_pair(c, locklocation);
+    if (fTry)
+    {
+        // try locks can not be waiting
+        now.second.ChangeWaitingToHeld();
+    }
+    // tid of the originating request
+    const uint64_t tid = getTid();
+    // If this is a blocking lock operation, we want to make sure that the locking order between 2 mutexes is consistent
+    // across the program
+    if (fTry)
+    {
+        // a try lock will either get it, or it wont. so just add it.
+        // if we dont get the lock this will be undone in the destructor of the read or write block
+        // for whichever critical section made this function call.
+        AddNewLock(now, tid);
```

Comment:
This limits the code structure.  I can't do this:
{
LOCK(A)
LOCK(B)
...
}

if (try_lock(B)) {}
else
{
LOCK(A);  // this should work because B is guaranteed not locked within here.  But it won't work based on this code which just assumes try_locks worked...
}

This is clearly an esoteric case.  But can it be handled easily?

Reply:
this will not fail for live deadlock detection because there is no live deadlock. 
this will not fail for lock order detection because neither the old nor the new code runs order detection on try locks

i will rework the comment in the code to make this more clear

Reply:
reworked